Adding availableKeys to ParseObject.State

natario1 opened the pull request:
- Adds a availableKeys property to ParseObject.state that includes keys that can be safely accessed. Fixes #595 so now there is consistency between complete and non-complete objects.

   While it seems to work on my machine, I have no deep understanding of the library. Nothing should be broken though. 

- Exposing `ParseObject.isDataAvailable(String key)`. Right now there are:

    -  `ParseObject.has(String key)`: alias to containsKey(key)
    - `ParseObject.containsKey(String key):` true if the server sent out data for this key

    Neither of these can be used to know if `ParseObject.get*` methods can be called safely. In some cases `has(key)` returns false but `get(key)` can be called without throwing. In some others `has(key)` returns false and `get(key)` will throw.

    This PR exposes `isDataAvailable(String key)` for partially fetched objects. When false, `get*` methods will throw. So with the PR:

   - `ParseObject.containsKey(String key)`: we have some data for this key from the server.
   - `ParseObject.isDataAvailable()`: the object is complete, all getters can be called.
   - `ParseObject.isDataAvailable(String key)`: the object might not be complete, but you can call get(key) safely.
    
  This makes sense to me.

- Fixes #597 adding support for nested selected keys. If there are some, nested objects will not be considered complete.

rogerhu commented:
Thanks.  Do you know how the iOS version handles this issue?

On Sat, Mar 11, 2017, 9:22 AM Mattia <notifications@github.com> wrote:

>
>    -
>
>    Adds a safeKeys property to ParseObject.state that includes keys that
>    can be safely accessed. Fixes #595
>    <https://github.com/ParsePlatform/Parse-SDK-Android/issues/595> so now
>    there is consistency between complete and non-complete objects.
>
>    While it seems to work on my machine, I have no deep understanding of
>    the library. Nothing should be broken though.
>    -
>
>    Exposing ParseObject.isDataAvailable(String key). Right now there are:
>
>    ParseObject.has(String key): alias to containsKey(key)
>    ParseObject.containsKey(String key): true if the server sent out data
>    for this key
>
>    Neither of these can be used to know if ParseObject.get* methods can
>    be called safely. In some cases has(key) returns false but get(key)
>    can be called without throwing. In some others has(key) returns false
>    and get(key) will throw.
>
> This PR exposes isDataAvailable(String key) for partially fetched
> objects. When false, get* methods will throw. So with the PR:
>
> ParseObject.has(String key): we have some data for this key from the
> server.
> ParseObject.containsKey(String key): same.
> ParseObject.isDataAvailable(): the object is complete, all getters can be
> called.
> ParseObject.isDataAvailable(String key): the object might not be
> complete, but you can call get(key) safely.
>
> This makes sense to me.
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/ParsePlatform/Parse-SDK-Android/pull/596
> Commit Summary
>
>    - Add safeKeys that can be safely accessed. Exposing
>    isDataAvailable(key)
>
> File Changes
>
>    - *M* Parse/src/main/java/com/parse/NetworkQueryController.java
>    <https://github.com/ParsePlatform/Parse-SDK-Android/pull/596/files#diff-0>
>    (2)
>    - *M* Parse/src/main/java/com/parse/ParseObject.java
>    <https://github.com/ParsePlatform/Parse-SDK-Android/pull/596/files#diff-1>
>    (39)
>
> Patch Links:
>
>    - https://github.com/ParsePlatform/Parse-SDK-Android/pull/596.patch
>    - https://github.com/ParsePlatform/Parse-SDK-Android/pull/596.diff
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/ParsePlatform/Parse-SDK-Android/pull/596>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAT8yWCWnGFMtnutVzSrsQT8kks6mniKks5rkthpgaJpZM4MaRF1>
> .
>

natario1 commented:
@rogerhu I don't know. As far as I know the JS sdk simply does not throw, so you can call get with any key, and if not present, you get undefined.

flovilmart commented:
What's the behavior on the iOS SDK? Should remove the throwing altogether?

rogerhu commented:
So iOS Parse has selectKeys() and findObjects() as part of [PFQuery] (https://github.com/ParsePlatform/Parse-SDK-iOS-OSX/blob/7a820b75c6726808475af9ce705053ff0e8cbd11/Parse/PFQuery.m) and [PFQueryState] (https://github.com/ParsePlatform/Parse-SDK-iOS-OSX/blob/7a820b75c6726808475af9ce705053ff0e8cbd11/Parse/Internal/Query/State/PFQueryState.m)

Is selectedKeys() the equivalent of safeKeys? (https://github.com/ParsePlatform/Parse-SDK-iOS-OSX/blob/7a820b75c6726808475af9ce705053ff0e8cbd11/Tests/Unit/QueryStateUnitTests.m)

natario1 commented:
@rogerhu we also have selectedKeys() as part of ParseQuery.State and that’s used to understand if the object should be considered ‘complete’. So if you select some keys all the objects will be incomplete. The issue is that when objects are incomplete and you call get* on a legit key, ParseObject will throw.

The PR adds safeKeys to ParseObject.State , meaning, ‘keys we don’t currently have in our data, but you should be able to call get() on without crashing the app’.

This is dirty and I would be OK with @flovilmart with simply removing the throwing.

rogerhu commented:
Where do we need to remove IllegalStateException throwing?  I think that makes more sense too to keep consistent with the other SDKs.

natario1 commented:
I have found how iOS handles this issue, [exactly here](https://github.com/ParsePlatform/Parse-SDK-iOS-OSX/blob/master/Parse/PFObject.m#L767).

There’s a `availableKeys` set, that holds actual "keys received from the server" **plus** "keys selected on the query".

This is practically the same to this PR (except that here the `safeKeys` Set means "keys selected on the query but not received from the server"). So this looks OK to me. What do you think?

rogerhu commented:
Should we call it availableKeys instead then?

rogerhu commented:
Does it deal with dotted notation as well?

natario1 commented:
Yes, I can refactor to give the same meaning it has in the iOS SDK. It’s a bit risky though, because in that case we must ensure that the `availableKeys` set is always consistent with the `serverData` map.

I don’t fully understand that part (server data is held in two different maps both in `ParseObject` and in `ParseObject.State`...). 

Yes, this implements dotted notation.

rogerhu commented:
I'm more curious if there's anywhere in the iOS code that deals with dotted notation issues as well...

natario1 commented:
@rogerhu it seems they don't, so they should have the #597 bug.

The relevant part is [here](https://github.com/ParsePlatform/Parse-SDK-iOS-OSX/blob/master/Parse/Internal/PFDecoder.m#L89): every decoded nested object is created as ‘complete’, which is wrong if some keys where selected.

The exact same line for the Android SDK is [here](https://github.com/ParsePlatform/Parse-SDK-Android/pull/596/files#diff-e30b7266f09ee255a64daa415de45d25L125) and this fixes it.

rogerhu commented on Parse/src/main/java/com/parse/ParseObject.java:
```diff
@@ -921,8 +939,34 @@ protected boolean visit(Object object) {
+          JSONArray nestedKeys = new JSONArray();
+          for (int i = 0; i < selectedKeys.length(); i++) {
+            String nestedKey = selectedKeys.getString(i);
+            if (nestedKey.startsWith(key+".")) nestedKeys.put(nestedKey.substring(key.length()+1));
```
What happena if this is multiple nested keys?  A.b.c.d.e?

natario1 replied:
At first (942-954) the current object will be given its own direct keys (without dots, e.g. b). 
Later (958+) we extract the child object subkeys (e.g. b.c.d.e). That is passed to decoder.decode(), which will pass back the JSON object to this method. It will get ‘b’ as a safe key, and will pass ‘c.d.e’... and so on. It goes this way with nested iterations, until there are no nested ParseObjects.

rogerhu replied:
Great thanks for the explanation

rogerhu replied:
Do we have a test for this part?

natario1 replied:
I have added two in ParseDecoderTest, I will add another one for the a.b situation.

rogerhu replied:
Thanks also the key + ”.” could be spaced out

rogerhu commented on Parse/src/test/java/com/parse/ParseObjectTest.java:
```diff
@@ -237,7 +238,16 @@ public void testGetUnavailable() {
     when(state.className()).thenReturn("TestObject");
     when(state.isComplete()).thenReturn(false);
+    ParseObject object = ParseObject.from(state);
+    object.get("foo");
```
Is this just an assertion that no exception will be raised?

natario1 replied:
No, this checks that the exception is raised (see the @Test annotation). Actually I have not changed this method, I have added testGetAvailableIfKeySafe() a few lines below, that tests safeKeys() .

rogerhu commented on Parse/src/main/java/com/parse/ParseObject.java:
```diff
@@ -578,33 +599,30 @@ public Void then(Task<Void> task) throws Exception {
    * @param isComplete
    *          {@code true} if this is all of the data on the server for the object.
    * @param decoder
-   *          Delegate for knowing how to decode the values in the JSON.
```
Was this line deleted accidentally?

natario1 replied:
Yes, good catch.

rogerhu commented on Parse/src/main/java/com/parse/NetworkQueryController.java:
```diff
@@ -134,9 +134,12 @@ public Integer then(Task<JSONObject> task) throws Exception {
       for (int i = 0; i < results.length(); ++i) {
         JSONObject data = results.getJSONObject(i);
-        T object = ParseObject.fromJSON(data, resultClassName, state.selectedKeys() == null);
+        if (isSubset) data.put(ParseObject.KEY_SELECTED_KEYS, selectedKeys);
```
Can you pass in selectedKeys into a method instead that accepts selectedKeys() instead?

https://github.com/ParsePlatform/Parse-SDK-iOS-OSX/blob/master/Parse/PFObject.m#L772-L774

It seems weird that the NetworkQueryController has access to this private key..

natario1 replied:
@rogerhu I get your point, I tried working directly with a Set, but for that we would need to change the signatures of ParseObject.fromJSON, ParseObject.mergeFromServer, and even worse ParseDecoder.decode, these are called from anywhere and an extra set of "selected keys" rarely makes sense.
The key is also used by ParseDecoder to get keys from the JSON...

Maybe we could add two static methods to ParseObject, `void insertSelectedKeys(JSONObject data, Set<String>)` , used by NetworkQueryController, and `boolean hasSelectedKeys(JSONObject data)` , used by ParseDecoder. What do you think?

natario1 replied:
Or I’ll just add another ParseObject.fromJSON that deals with this.

rogerhu replied:
Yep :)

rogerhu commented on Parse/src/main/java/com/parse/ParseDecoder.java:
```diff
@@ -122,7 +122,7 @@ public Object decode(Object object) {
 
     if (typeString.equals("Object")) {
-      return ParseObject.fromJSON(jsonObject, null, true, this);
+      return ParseObject.fromJSON(jsonObject, null, !jsonObject.has(ParseObject.KEY_SELECTED_KEYS), this);
```
it seems to me that fromJSON() should figure out whether things are complete instead of here right?

Just trying to get KEY_SELECTED_KEYS out of being a public static variable..

rogerhu commented:
Thanks for working on this bug/issue..quite complicated to fix and appreciate the effort.  Most of my feedback are minor issues in general.

natario1 commented on Parse/src/main/java/com/parse/ParseObject.java:
```diff
@@ -921,8 +940,34 @@ protected boolean visit(Object object) {
+            ((JSONObject) value).put(KEY_SELECTED_KEYS, nestedKeys);
+          }
+        }
         Object decodedObject = decoder.decode(value);
```
Well we need a fromJSON signature that does not take selectedKeys() .. because ParseDecoder will just have a JSON and is not aware of the selected keys bundled in the JSON (I made KEY_SELECTED_KEYS private as you said), [see](https://github.com/ParsePlatform/Parse-SDK-Android/pull/596/files#diff-e30b7266f09ee255a64daa415de45d25L125).

From this point on, bundling the keys in the JSON allows us to not change all the other signatures, and abstract this internal stuff from ParseDecoder. It is also used in Local data store encoding/decoding, see ParseObject.toRest() / fromREST()

rogerhu replied:
Gotcha forgot about local store too

natario1 commented on Parse/src/main/java/com/parse/ParseObject.java:
```diff
@@ -98,6 +99,7 @@ private static ParseObjectSubclassingController getSubclassingController() {
       private long updatedAt = -1;
       private boolean isComplete;
+      private Set<String> safeKeys = new HashSet<>();
       /* package */ Map<String, Object> serverData = new HashMap<>();
```
Haha yes. Actually the server does not return these, we have them because of query.selectKeys(), but you got the point. Annoying bug.

rogerhu commented:
Lgtm thanks

flovilmart commented:
Can we rename those to availableKeys?

rogerhu commented:
@natario1 how easy would it be to change the construct to be availableKeys?  The main thing I think is this function here: https://github.com/natario1/Parse-SDK-Android/blob/fcf24e3af94f0321d80d399c418a325e7e3da90f/Parse/src/main/java/com/parse/ParseObject.java#L168-L174

natario1 commented:
Ok guys, reverted that logic and renamed to availableKeys() .

Since this has become something serious, the next design question would be if we should add the set also to `ParseObject` (not only `ParseObject.State`). As far as I understand the `ParseObject.State` is synced with the server, while `ParseObject` members do also trace dirty changes.

So, for example, you have the `ParseObject.State.serverData` map and the `ParseObject.estimatedData` map. The latter is modified by put operations. It appears they are synced together after save operations.

So, now that we have `ParseObject.State.availableKeys`, should we also add `ParseObject.estimatedKeys`?

- Pros: would allow to change [this line](https://github.com/natario1/Parse-SDK-Android/blob/selectkeys/Parse/src/main/java/com/parse/ParseObject.java#L3446) from

        public boolean isDataAvailable(String key) {
          synchronized (mutex) {
            // Fallback to estimatedData to include dirty changes.
           return isDataAvailable() || state.availableKeys().contains(key) || estimatedData.containsKey(key);
          }
        }

   to a much cleaner

        public boolean isDataAvailable(String key) {
          synchronized (mutex) {
           return isDataAvailable() || estimatedKeys.contains(key);
          }
        }

- Pros: consistency with `estimatedData` and with the iOS SDK implementation of isDataAvailable

- Cons: adding a useless Set for every ParseObject (it will differ from `estimatedData.keySet()` only when the object comes from a Query with selected keys...).

- Cons: must be 100% sure that `estimatedKeys` follows operation onto `estimatedData` or we will throw when we shouldn't. Changing the implementation of an important method (`isDataAvailable(key)`) and I don’t think I have enough understanding of the class.

What do you think @rogerhu @flovilmart ? I think it can be part of a future PR ...

natario1 commented:
Just to clarify, the current implementation (with this PR) of `isDataAvailable(key)` **does** follow dirty changes. This works. We just have to check estimated data.

flovilmart commented:
If it works, don't change the internal yet :)

rogerhu commented:
Good enough for now. :). Thanks!

natario1 commented:
When are you planning to release next version @rogerhu ?

flovilmart commented:
Snapshots are released with each merge. That should be good to test no?

rogerhu commented:
Working also on Bintray access so I can deploy the v14.0.0 version currently on snapshot too. :)

natario1 commented:
Nice, no hurry. By the way, we are pushing 1.14.1 to Sonatype, is that a typo? (should be 1.14.0)

rogerhu commented:
Yeah my fail.  :)  It's 1.14.1-SNAPSHOT (I'll fix it to 1.14.0)

rogerhu commented on Parse/src/main/java/com/parse/NetworkQueryController.java:
```diff
@@ -136,7 +136,7 @@ public Integer then(Task<JSONObject> task) throws Exception {
       for (int i = 0; i < results.length(); ++i) {
         JSONObject data = results.getJSONObject(i);
-        T object = ParseObject.fromJSON(data, resultClassName, state.selectedKeys() == null);
+        T object = ParseObject.fromJSON(data, resultClassName, ParseDecoder.get(), state.selectedKeys());
```
@natario1 do you remember why we need to pass in ParseDecoder.get() here?

natario1 replied:
@rogerhu it’s just an instance of ParseDecoder.

I had to change the signatures of ParseObject.fromJSON and this seemed legit. ParseDecoder was present in the other signature (used [here](https://github.com/ParsePlatform/Parse-SDK-Android/blob/master/Parse/src/main/java/com/parse/ParseDecoder.java#L125) by ParseDecoder itself, passing `this`), so I added it here as well for consistency.

Since ParseDecoder is a singleton we could just drop it from both signatures, but I guess original authors wanted to keep the chance of passing a different decoder. I don’t know.
